clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -6183,7 +6212,78 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
   O << "} // namespace detail\n";
   O << "} // namespace _V1\n";
   O << "} // namespace sycl\n";
-  O << "\n";
+
+  unsigned ShimCounter = 1;
+  int FreeFunctionCount = 0;
+  for (const KernelDesc &K : KernelDescs) {
+    if (!isFreeFunction(S, K.SyclKernel))
+      continue;
+
+    ++FreeFunctionCount;
+    O << "\n// Definition of " << K.Name << " as a free function kernel\n";
+    // Currently all parameters are scalars or pointers. When more
+    // complex types are supported, use the FwdDeclEmitter to generate forward
+    // type declarations.
+    if (K.SyclKernel->getLanguageLinkage() == CLanguageLinkage)
```

Comment:
Do you need to consider namespaces?

Reply:
Can you also explain why you need to print the non-templated version?

Reply:
Namespaces are not currently supported.

Let me try to explain what needs to be placed in the integration header.

Consider a free function declared and instantiated twice as follows:
```
template <int N, bool AIsVector>
SYCL_EXTERNAL SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(
    (ext::oneapi::experimental::
         single_task_kernel)) void simd_f(int ValuesToTrySize,
                                          int *ValuesToTryPtr) {}
// Explicit instantiations
template void simd_f<16, true>(int, int *);
template void simd_f<8, true>(int, int *);
```
The integration header needs these declarations:
```
// Forward declaration of templated free function
template <int, bool> void simd_f(int, int *);
// Shim declarations
static constexpr auto __sycl_shim1() { return (void (*)(int, int *))simd_f<16, true>; }
static constexpr auto __sycl_shim2() { return (void (*)(int, int *))simd_f<8, true>; } 
```

The SYCL kernel functions produced for the above usage will print like this:
`template <> void simd_f(int, int *);`

This is not directly useful. So we either have to create FunctionDecls that using standard printers will produce what is needed, or try and do text manipulation on bits and pieces of the SYCL kernel FunctionDecl.

Reply:
>> The SYCL kernel functions produced for the above usage will print like this:
template <> void simd_f(int, int *);

This seems odd to me. Have you tried playing with options in `PrintingPolicy` which may control how templates are printed? I do not know off the top of my head which option controls this or if an option for this even exists but I suggest looking into it. I see various options like `SuppressDefaultTemplateArgs`, etc which may control this.

Reply:
None of the Policy switches helped with the functionality needed for free functions. Currently, the various function declarations are produced individually. In a future PR, with support for namespaces and more complex types, a new or modified FwdDeclprinter will be implemented.

Reply:
Hmm I see. Did you try printing SYCLKernel? I tried the example you shared and this is what printed for of your tests below - 

```
template<> void ff_3<double>(double *ptr, double start, double end) {
    for (int i = start; i <= end; i++)
        ptr[i] = end;
}
```
Is this not correct?